broadway_api/handlers/worker.py

```diff
@@ -17,32 +17,43 @@
 class WorkerRegisterHandler(BaseAPIHandler):
     @authenticate_cluster_token
     @schema.validate(
+        on_empty_404=True,
+        input_schema={
+            "type": "object",
+            "properties": {"hostname": {"type": "string"}},
+            "required": ["hostname"],
+            "additionalProperties": False,
+        },
         output_schema={
             "type": "object",
-            "properties": {
-                "worker_id": {"type": "string"},
-                "heartbeat": {"type": "number"},
-            },
-            "required": ["worker_id"],
+            "properties": {"heartbeat": {"type": "number"}},
+            "required": ["heartbeat"],
             "additionalProperties": False,
-        }
+        },
     )
-    def get(self, *args, **kwargs):
+    def post(self, *args, **kwargs):
+        worker_id = kwargs.get("worker_id")
+        hostname = self.body.get("hostname")
+
         worker_node_dao = daos.WorkerNodeDao(self.settings)
+
+        dup = worker_node_dao.find_by_worker_id(worker_id)
+
+        if dup is not None and dup.is_alive:
+            msg = "worker id '{}' already exists".format(worker_id)
+            logger.critical(msg)
+            self.abort({"message": msg}, status=400)
+            return
+
         worker_node = models.WorkerNode(
-            hostname=kwargs.get("hostname") if len(args) == 0 else args[0],
-            last_seen=get_time(),
-            is_alive=True,
+            worker_id=worker_id, hostname=hostname, last_seen=get_time(), is_alive=True
         )
-        worker_id = str(worker_node_dao.insert(worker_node).inserted_id)
 
-        logger.info(
-            "worker '{}' joined as id '{}'".format(worker_node.hostname, worker_id)
-        )
-        return {
-            "worker_id": worker_id,
-            "heartbeat": self.get_config()["HEARTBEAT_INTERVAL"],
-        }
+        worker_node_dao.update(worker_node)
+
+        logger.info("worker '{}' joined as id '{}'".format(hostname, worker_id))
```

Comment:
the log message should indicate the difference between a re-register and addition.